<a href="https://colab.research.google.com/github/aishwary247/Deep-Learning/blob/main/Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
!pip install unidecode

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.5/235.5 kB 4.5 MB/s eta 0:00:00


In [ ]:
df = pd.read_csv('/content/Corona_train.csv')
df_test = pd.read_csv("/content/Corona_validation.csv")
df.head()

ID Sentiment                                        CoronaTweet
0  22979  Positive  I see all kinds of academics already whipping ...
1   9880  Negative  @HenrySmithUK can you raise with Boris please ...
2  35761  Negative  It s a confusing odd time for the shopping pub...
3  37968  Positive  Blog Summary: The Impact of COVID-19 on the Ca...
4  19709   Neutral  ??????? ??????? ???\r\r\nWaiting in a long Que...

# Text Preprocessing

In [ ]:
df_original = df.copy()
df_test_original = df_test.copy()

In [ ]:
import nltk
nltk.download('punkt')
from nltk.corpus import stopwords
nltk.download('stopwords')
stwords = stopwords.words('english')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
text = "Hey covid19"
text = re.sub('[^a-zA-Z\s]', '', text)
text

'Hey covid'

In [ ]:
# import re
# from nltk.stem.porter import PorterStemmer
# from unidecode import unidecode


# # Removing URL's
# def remove_url(text):
#   pattern = re.compile(r'https?://\S+|www\.\S+')
#   return pattern.sub(r'',text)


# # Define a function for text cleaning
# def clean_text(text):
#     if pd.notnull(text) and isinstance(text, str):
#         text = text.lower()
#         text = re.sub('<br />', '', text)
#         text = re.sub('\n', '', text)
#         # text = re.sub('[^\w\s]', '', text)
#         text = re.sub('[^a-zA-Z\s]', '', text)
#     return text

# ps = PorterStemmer()

# def stem(text):
#   y=[]
#   for i in text.split() :
#     y.append(ps.stem(i))

#   return " ".join(y)

# def remove_stopwords(text):
#     stop_words = set(stopwords.words('english'))
#     words = nltk.word_tokenize(text)
#     filtered_words = [word for word in words if word.lower() not in stop_words]
#     return ' '.join(filtered_words)


# def strip_accents(text):
#     return unidecode(text)

In [ ]:
import re
from nltk.stem.porter import PorterStemmer
from unidecode import unidecode
def preprocess_text(text):
  """
  This function preprocesses text data for further analysis.

  Args:
      text (str): The text string to be preprocessed.

  Returns:
      str: The preprocessed text string.
  """
  if pd.notnull(text) and isinstance(text, str):

    pattern = re.compile(r'https?://\S+|www\.\S+')
    text = pattern.sub(r'',text)
    # Lowercase
    text = text.lower()

    # Remove HTML tags and newlines
    text = re.sub('<br />', '', text)
    text = re.sub('\n', '', text)

    text = re.sub('[^a-zA-Z\s]', '', text)


    # Stemming
    ps = PorterStemmer()
    text = " ".join([ps.stem(i) for i in text.split()])

    # Remove stop words
    stop_words = set(stopwords.words('english'))
    words = nltk.word_tokenize(text)
    filtered_words = [word for word in words if word.lower() not in stop_words]
    text = ' '.join(filtered_words)

    # Remove accents
    text = unidecode(text)

    return text


In [ ]:
import time

start = time.time()

df["CoronaTweet"] = df["CoronaTweet"].apply(preprocess_text)

time1 = time.time()-start
df_test["CoronaTweet"] = df_test["CoronaTweet"].apply(preprocess_text)

time2 = time.time()-start - time1

print(f"Time1: {time1} Time2: {time2}")

Time1: 41.435137033462524 Time2: 3.137101173400879


In [ ]:
# df["CoronaTweet"]= df["CoronaTweet"].apply(remove_url)
# df["CoronaTweet"]= df["CoronaTweet"].apply(clean_text)
# df["CoronaTweet"]= df["CoronaTweet"].apply(stem)
# df["CoronaTweet"]= df["CoronaTweet"].apply(remove_stopwords)
# df["CoronaTweet"]= df["CoronaTweet"].apply(strip_accents)

In [ ]:
df

ID Sentiment                                        CoronaTweet
0      22979  Positive  see kind academ alreadi whip covid relat proje...
1       9880  Negative  henrysmithuk rais bori pleas supermarket contr...
2      35761  Negative  confus odd time shop public store close flight...
3      37968  Positive  blog summari impact covid canadian residenti h...
4      19709   Neutral  wait long queue enter supermarket final saw en...
...      ...       ...                                                ...
37859  20253   Neutral  doubl tap rememb employ follow walletgyd tip l...
37860  38926  Positive  struggl understand whi supermarket food shop a...
37861  35889  Positive  anyon interest checklist consum implic covid p...
37862  40641  Positive  agre hypothesi indian societi might chang fore...
37863  26063   Neutral  richmond local boutiqu thrift store search new...

[37864 rows x 3 columns]

In [ ]:
df_test

ID Sentiment                                        CoronaTweet
0      7184  Negative  reflect consum behaviour last week made thi li...
1     36363  Negative  know everyon get stir crazi befor go watch thi...
2     10423  Negative  seen ga price thi low sinc first start drive l...
3      6409   Neutral  onli batmeat left supermarket shelv coronaviru...
4      7015   Neutral  along health worker need appreci truck driver ...
...     ...       ...                                                ...
3288  13685   Neutral  day selfisol run quick errand buy supermarket ...
3289  17905  Positive  tesco think dure covid pandem bag onlin shop o...
3290  40331   Neutral           key insight effect covid consum behavior
3291   1495   Neutral  score haul groceri store tonight hoard coronav...
3292   3190  Positive  moral justifi busi rais price good demand grow...

[3293 rows x 3 columns]

In [ ]:
# # Function to remove words with numbers, excluding cases like "covid19"
# def remove_numerical_words(text):
#     return ' '.join(word for word in text.split() if not (re.search(r'\d', word) and not re.search(r'[\W_]*\d*covid[\W_]*\d*', word, flags=re.IGNORECASE)))

# def replace_covid_words(text):
#     return re.sub(r'[\W_]*\d*covid[\W_]*\d*', 'covid', text, flags=re.IGNORECASE)
# df['CoronaTweet']= df['CoronaTweet'].apply(remove_numerical_words)
# df['CoronaTweet']= df['CoronaTweet'].apply(replace_covid_words)
# df


In [ ]:
def get_vocabulary(text_col):
  """
  This function finds the unique vocabulary (words) in a DataFrame column.

  Args:
      text_col (pd.Series): The DataFrame column containing text data.

  Returns:
      set: A set containing the unique words found in the column.
  """
  # Combine all text into a single string (optional for short columns)
  all_text = " ".join(text_col.tolist())

  # Tokenize the text (split into words)
  words = all_text.lower().split()  # Convert to lowercase for case-insensitivity

  # Remove empty strings or punctuation (optional)
  # words = [word for word in words if word.isalpha()]  # Only keep alphabetic words

  # Find unique words
  vocabulary = set(words)

  return vocabulary

# Get the vocabulary of the 'text' column
vocabulary = get_vocabulary(df['CoronaTweet'])

# Print the vocabulary
print(vocabulary)

{'gravit', 'ayuk', 'amygavin', 'dramaass', 'allah', 'gailafonso', 'router', 'significantdisrupt', 'berkwatch', 'lloydblankfein', 'impl', 'swear', 'absenc', 'ibrahim', 'thati', 'maldivianaero', 'maskscov', 'rospotrebnadsor', 'wtrg', 'magic', 'shortrun', 'headband', 'albania', 'utilityweek', 'todayrollingston', 'peterschiff', 'precaucao', 'turnov', 'baru', 'mumbaincp', 'mid', 'boblouka', 'varic', 'khutso', 'postdisast', 'millionair', 'directtoconsum', 'motiongraph', 'blackop', 'cheapga', 'adgulli', 'groundbreak', 'cambrial', 'pleaseeeeeee', 'sauer', 'albermarl', 'yourhom', 'like', 'substanc', 'bludgeon', 'deliberatli', 'dontneedatherapist', 'happi', 'uneth', 'drgregpoland', 'ogmundson', 'grande', 'soldoitofvitamin', 'nice', 'manchesterc', 'pilgrim', 'milyon', 'flashi', 'fivemonth', 'selon', 'incap', 'covidaustralia', 'titl', 'realoverheardla', 'luckycharm', 'frenchfri', 'streetadvoc', 'vicksiern', 'germx', 'theresalwayson', 'mddc', 'gersheim', 'imposit', 'maghnayeh', 'affiliatemarket', '

In [ ]:
vocubulary_size = len(vocabulary)
print(vocubulary_size)

44100


## Using Count Vectoriser and Tfidf

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
max_features = 10000
cv = CountVectorizer(max_features = max_features ,stop_words='english') # can keep max features as a hyperparameter
tfidf_vectorizer = TfidfVectorizer(max_features=max_features, stop_words='english')


In [ ]:
vector = cv.fit_transform(df['CoronaTweet']).toarray()
vector_test = cv.transform(df_test['CoronaTweet']).toarray()

vector_tf = tfidf_vectorizer.fit_transform(df['CoronaTweet']).toarray()
vector_tf_test = tfidf_vectorizer.transform(df_test['CoronaTweet']).toarray()

vector.shape, vector_tf.shape, vector_test.shape, vector_tf_test.shape

((37864, 10000), (37864, 10000), (3293, 10000), (3293, 10000))

In [ ]:
arr = cv.get_feature_names_out()
print(arr[9900:10000])

['wrap' 'wreak' 'wreck' 'wrench' 'wrestl' 'wrestlemania' 'write' 'writer'
 'writingcommnun' 'writingcommun' 'written' 'wrong' 'wrote' 'ws' 'wsj'
 'wtf' 'wth' 'wti' 'wtop' 'wuflu' 'wuhan' 'wuhancoronaviru' 'wuhanviru'
 'wusa' 'wv' 'ww' 'wwd' 'wwe' 'wwgwga' 'wwii' 'xauusd' 'xbox' 'xd' 'xma'
 'xx' 'ya' 'yahoo' 'yahoofin' 'yahoonew' 'yall' 'yangon' 'yard' 'yay' 'ye'
 'yeah' 'year' 'yearli' 'yearold' 'yeast' 'yeg' 'yell' 'yellow' 'yelp'
 'yemen' 'yep' 'yesterday' 'yield' 'yike' 'ynwa' 'yo' 'yoga' 'yoghurt'
 'yogi' 'yomequedoencasa' 'york' 'yorker' 'yorkshir' 'yotpo' 'youd'
 'yougov' 'youll' 'youllneverwalkalon' 'young' 'younger'
 'yoursafetyismysafeti' 'yourselv' 'youth' 'youthl' 'youtub' 'youv' 'yoy'
 'yr' 'yrold' 'yu' 'yummi' 'yup' 'yvr' 'yyc' 'zambia' 'zealand' 'zero'
 'zimbabw' 'zinc' 'zomatoin' 'zombi' 'zombieland' 'zone' 'zoom'
 'zoomupyourlif' 'zoonot']


In [ ]:
df.shape[0]

37864

In [ ]:
df.Sentiment.unique()

array(['Positive', 'Negative', 'Neutral'], dtype=object)

In [ ]:
# m = df.shape[0]
# y = np.zeros((m))
# for i in range(m):
#   if df["Sentiment"][i] == "Positive" :
#     y[i] = 0
#   elif df["Sentiment"][i] == "Neutral" :
#     y[i] = 1
#   else :
#     y[i] = 2

# print(y)
sentiment_mapping = {"Positive": 0, "Neutral": 1, "Negative": 2}
y = np.array([sentiment_mapping[s] for s in df["Sentiment"]])
print(y)

[0 2 2 ... 0 0 1]


In [ ]:
y_test = np.array([sentiment_mapping[s] for s in df_test["Sentiment"]])
y_test

array([2, 2, 2, ..., 1, 1, 0])

#Implementing Naive Bayes

In [ ]:
# # vector, y
# m = 37864

# sigmay0 = 0
# sigmay1 = 0
# sigmay2 = 0
# for ele in y :
#   if ele == 0:
#     sigmay0 += 1
#   elif ele == 1:
#     sigmay1 += 1
#   else :
#     sigmay2 += 1
sentiment_counts = np.bincount(y)

# Extract counts for each sentiment class (0, 1, 2)
sigmay0 = sentiment_counts[0] if len(sentiment_counts) > 0 else 0
sigmay1 = sentiment_counts[1] if len(sentiment_counts) > 1 else 0
sigmay2 = sentiment_counts[2] if len(sentiment_counts) > 2 else 0

# Print the counts
print("Number of zeros:", sigmay0)
print("Number of ones:", sigmay1)
print("Number of twos:", sigmay2)
print(sigmay0+sigmay1+sigmay2)

Number of zeros: 16602
Number of ones: 7096
Number of twos: 14166
37864


In [ ]:
m, n = vector.shape
phi0 = np.zeros((n))
phi1 = np.zeros((n))
phi2 = np.zeros((n))


# for k in range(n):
#   num0 = 0
#   num1 = 0
#   num2 = 0

#   for i in range(m):
#     if vector[i][k] == 1 and y[i] == 0 :
#       num0 += 1
#     elif vector[i][k] == 1 and y[i] == 1 :
#       num1 += 1
#     elif vector[i][k] == 1 and y[i] == 2 :
#       num2 += 1
#   phi0[k] =  (num0 + 1)/(sigmay0+3)
#   phi1[k] =  (num1 + 1)/(sigmay1+3)
#   phi2[k] =  (num2 + 1)/(sigmay2+3)



for k in range(n):
    num0 = np.sum((vector[:, k] == 1) & (y == 0))
    num1 = np.sum((vector[:, k] == 1) & (y == 1))
    num2 = np.sum((vector[:, k] == 1) & (y == 2))

    phi0[k] = (num0 + 1) / (sigmay0 + 3)
    phi1[k] = (num1 + 1) / (sigmay1 + 3)
    phi2[k] = (num2 + 1) / (sigmay2 + 3)

In [ ]:
phiy = np.array((sigmay0/m, sigmay1/m, sigmay2/m))
phiy

array([0.43846398, 0.18740756, 0.37412846])

In [ ]:
# Baic implementation with loops not time efficient
#def sigmap( X , y ):
#   if(y==0):
#     pdt = 1
#     for i in range(X.shape[0]):
#       if X[i] == 0 :
#         pdt *=(1-phi0[i])
#       else :
#         pdt *= phi0[i]
#     return pdt

#   elif(y==1):
#     pdt = 1
#     for i in range(X.shape[0]):
#       if X[i] == 0 :
#         pdt *=(1-phi1[i])
#       else :
#         pdt *= phi1[i]
#     return pdt

#   else:
#     pdt = 1
#     for i in range(X.shape[0]):
#       if X[i] == 0 :
#         pdt *=(1-phi2[i])
#       else :
#         pdt *= phi2[i]
#     return pdt

# def probab( y , X):
#   return (sigmap(X,y)*phiy[y])/(sigmap(X,0)*phiy[0]+sigmap(X,1)*phiy[1]+sigmap(X,2)*phiy[2])

# def predict(X):
#   probabilities = [probab(0, X), probab(1, X), probab(2, X)]

#   if max(probabilities) == probabilities[0]:
#       return 0
#   elif max(probabilities) == probabilities[1]:
#       return 1
#   else:
#        return 2



In [ ]:
# import time
# start = time.time()
# print(predict(vector[0]))
# time1 = time.time()-start
# print(time1*30000)

In [ ]:
# using Numpy array operations to improve time efficiency
def sigmap(X, phi):
    return np.prod(np.where(X == 0, 1 - phi, phi))

def probab(y, X, phi, phiy):
    numerator = sigmap(X, phi[y]) * phiy[y]
    denominator = np.sum([sigmap(X, phi[i]) * phiy[i] for i in range(3)])
    return numerator / denominator

def predict(X, phi, phiy):
    probabilities = [probab(i, X, phi, phiy) for i in range(3)]
    return np.argmax(probabilities)



In [ ]:
phi = [phi0,phi1,phi2]

In [ ]:
import time
start = time.time()
print(predict(vector[0],phi,phiy))
time1 = time.time()-start
print(time1*30000)

0
93.7342643737793


In [ ]:
y_pred = np.zeros(m)
for i in range(m):
  y_pred[i] = predict(vector[i],phi,phiy)

In [ ]:
y_pred

array([0., 2., 2., ..., 1., 1., 1.])

In [ ]:
num_rows = vector_test.shape[0]
num_cols = vector_test.shape[1]
y_test_pred = np.zeros(num_rows)
for i in range(num_rows):
  y_test_pred[i] = predict(vector_test[i],phi,phiy)
y_test_pred

array([2., 2., 2., ..., 1., 1., 0.])

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
accuracy = accuracy_score(y, y_pred)

print(f'Accuracy: {accuracy:.4f}')

# Confusion matrix
conf_matrix = confusion_matrix(y, y_pred)
print('\nConfusion Matrix:')
print(conf_matrix)

# Classification report
class_report = classification_report(y, y_pred)
print('\nClassification Report:')
print(class_report)

test_accuracy = accuracy_score(y_test, y_test_pred)

print(f'Test_Accuracy: {test_accuracy:.4f}')

# Confusion matrix
conf_matrix_test = confusion_matrix(y_test, y_test_pred)
print('\nConfusion Matrix:')
print(conf_matrix_test)

# Classification report
class_report_test = classification_report(y_test, y_test_pred)
print('\nClassification Report:')
print(class_report_test)

Accuracy: 0.7695

Confusion Matrix:
[[13074  1334  2194]
 [ 1036  5211   849]
 [ 2140  1173 10853]]

Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.79      0.80     16602
           1       0.68      0.73      0.70      7096
           2       0.78      0.77      0.77     14166

    accuracy                           0.77     37864
   macro avg       0.75      0.76      0.76     37864
weighted avg       0.77      0.77      0.77     37864

/n/n
Accuracy: 0.7148

Confusion Matrix:
[[1059  138  247]
 [ 143  390   84]
 [ 220  107  905]]

Classification Report:
              precision    recall  f1-score   support

           0       0.74      0.73      0.74      1444
           1       0.61      0.63      0.62       617
           2       0.73      0.73      0.73      1232

    accuracy                           0.71      3293
   macro avg       0.70      0.70      0.70      3293
weighted avg       0.72      0.71      0.72      

In [ ]:

from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

def evaluate_model(y_true, y_pred):
  """
  This function evaluates a classification model by calculating accuracy, confusion matrix, and classification report.

  Args:
      y_true (np.array): The true target labels.
      y_pred (np.array): The predicted labels by the model.
  """
  # Accuracy
  accuracy = accuracy_score(y_true, y_pred)
  print(f'Accuracy: {accuracy:.4f}')

  # Confusion Matrix
  conf_matrix = confusion_matrix(y_true, y_pred)
  print('\nConfusion Matrix:')
  print(conf_matrix)

  # Classification Report
  class_report = classification_report(y_true, y_pred)
  print('\nClassification Report:')
  print(class_report)


# Naive Bayes using sklearn

In [ ]:
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB()
model.fit(vector,y)

MultinomialNB()

In [ ]:
y_pred = model.predict(vector)
y_pred_test = model.predict(vector_test)
print("Train Data: ")
evaluate_model(y, y_pred)

print("Test Data: ")
evaluate_model(y_test, y_pred_test)

Train Data: 
Accuracy: 0.7565

Confusion Matrix:
[[13204  1150  2248]
 [ 1554  4292  1250]
 [ 2097   922 11147]]

Classification Report:
              precision    recall  f1-score   support

           0       0.78      0.80      0.79     16602
           1       0.67      0.60      0.64      7096
           2       0.76      0.79      0.77     14166

    accuracy                           0.76     37864
   macro avg       0.74      0.73      0.73     37864
weighted avg       0.75      0.76      0.76     37864

Test Data: 
Accuracy: 0.6978

Confusion Matrix:
[[1078  114  252]
 [ 198  290  129]
 [ 203   99  930]]

Classification Report:
              precision    recall  f1-score   support

           0       0.73      0.75      0.74      1444
           1       0.58      0.47      0.52       617
           2       0.71      0.75      0.73      1232

    accuracy                           0.70      3293
   macro avg       0.67      0.66      0.66      3293
weighted avg       0.69      

## Using Tfidf

In [ ]:
model2 = MultinomialNB()
model2.fit(vector_tf,y)

MultinomialNB()

In [ ]:
y_pred = model.predict(vector)
y_pred_test = model.predict(vector_test)
print("Train Data: ")
evaluate_model(y, y_pred)

print("Test Data: ")
evaluate_model(y_test, y_pred_test)

Train Data: 
Accuracy: 0.7565

Confusion Matrix:
[[13204  1150  2248]
 [ 1554  4292  1250]
 [ 2097   922 11147]]

Classification Report:
              precision    recall  f1-score   support

           0       0.78      0.80      0.79     16602
           1       0.67      0.60      0.64      7096
           2       0.76      0.79      0.77     14166

    accuracy                           0.76     37864
   macro avg       0.74      0.73      0.73     37864
weighted avg       0.75      0.76      0.76     37864

Test Data: 
Accuracy: 0.6978

Confusion Matrix:
[[1078  114  252]
 [ 198  290  129]
 [ 203   99  930]]

Classification Report:
              precision    recall  f1-score   support

           0       0.73      0.75      0.74      1444
           1       0.58      0.47      0.52       617
           2       0.71      0.75      0.73      1232

    accuracy                           0.70      3293
   macro avg       0.67      0.66      0.66      3293
weighted avg       0.69      

**Remark:** We don't get much of a difference in accuracy using tfidf or count vectorizer

# Using RNN'S

In [ ]:
from keras.preprocessing.text import Tokenizer
top_words = 10000
tokenizer = Tokenizer(num_words = top_words)
tokenizer.fit_on_texts(df['CoronaTweet'])

In [ ]:
len(tokenizer.word_index)

44100

In [ ]:
sequences = tokenizer.texts_to_sequences(df['CoronaTweet'])
max_length = max(len(seq) for seq in sequences)
print(max_length)

37


In [ ]:
test_seqs = tokenizer.texts_to_sequences(df_test['CoronaTweet'])

In [ ]:
max_index = max(max(seq) for seq in sequences if len(seq) )  # Consider non-empty sequences
print(max_index)

9999


In [ ]:
lengths = [len(seq) for seq in sequences]

mean_length = np.mean(lengths)
median_length = np.median(lengths)
std_dev_length = np.std(lengths)

print("Mean Length:", mean_length)
print("Median Length:", median_length)
print("Standard Deviation:", std_dev_length)

Mean Length: 17.495246144094654
Median Length: 18.0
Standard Deviation: 6.563958166163938


In [ ]:
from keras.utils import pad_sequences
max_length = 25
sequences = pad_sequences(sequences, padding='post', maxlen=max_length)
sequences

array([[  51,  453, 4847, ...,   43,  127,   18],
       [ 323, 1649,   52, ...,  150,    0,    0],
       [1917, 2537,   20, ...,    0,    0,    0],
       ...,
       [ 222,  415, 4744, ...,    0,    0,    0],
       [ 669, 9982, 1121, ...,    0,    0,    0],
       [5540,   61, 4854, ...,    0,    0,    0]], dtype=int32)

In [ ]:

test_seqs = pad_sequences(test_seqs, padding='post', maxlen=max_length)
test_seqs.shape

(3293, 25)

In [ ]:
sequences.shape

(37864, 25)

In [ ]:
from keras import Sequential
from keras.layers import Dense,SimpleRNN,Embedding,Flatten
model = Sequential()
model.add(Embedding(10000,output_dim=10,input_length=25))
model.add(SimpleRNN(32,return_sequences=False))
model.add(Dense(3, activation='softmax'))

model.summary()

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 25, 10)            100000    
                                                                 
 simple_rnn (SimpleRNN)      (None, 32)                1376      
                                                                 
 dense (Dense)               (None, 3)                 99        
                                                                 
Total params: 101475 (396.39 KB)
Trainable params: 101475 (396.39 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 25, 10)            100000    
                                                              

In [ ]:
y.shape, y_test.shape

((37864,), (3293,))

In [ ]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['acc'])
history = model.fit(sequences, y , epochs=5, validation_data=(test_seqs,y_test))

Epoch 1/5
1184/1184 [==============================] - 14s 10ms/step - loss: 0.8056 - acc: 0.6439 - val_loss: 0.6375 - val_acc: 0.7561
Epoch 2/5
1184/1184 [==============================] - 11s 10ms/step - loss: 0.5166 - acc: 0.8092 - val_loss: 0.5596 - val_acc: 0.7853
Epoch 3/5
1184/1184 [==============================] - 12s 10ms/step - loss: 0.4010 - acc: 0.8587 - val_loss: 0.5747 - val_acc: 0.7944
Epoch 4/5
1184/1184 [==============================] - 12s 10ms/step - loss: 0.3285 - acc: 0.8880 - val_loss: 0.6196 - val_acc: 0.7823
Epoch 5/5
1184/1184 [==============================] - 12s 10ms/step - loss: 0.2693 - acc: 0.9099 - val_loss: 0.6945 - val_acc: 0.7786


## Trying to reduce overfitting by experimentation

In [ ]:
from keras.preprocessing.text import Tokenizer
top_words = 3000
tokenizer = Tokenizer(num_words = top_words, oov_token='<nothing>')
tokenizer.fit_on_texts(df['CoronaTweet'])

sequences = tokenizer.texts_to_sequences(df['CoronaTweet'])
test_seqs = tokenizer.texts_to_sequences(df_test['CoronaTweet'])

max_length = 25
sequences = pad_sequences(sequences, padding='post', maxlen=max_length)

test_seqs = pad_sequences(test_seqs, padding='post', maxlen=max_length)

In [ ]:
from keras import Sequential
from keras.layers import Dense,SimpleRNN,Embedding,Flatten
from keras.regularizers import L1, L2

model = Sequential()
model.add(Embedding(3000,output_dim=10,input_length=25))
model.add(SimpleRNN(32,return_sequences=False))
model.add(Dense(3, activation='softmax')) #,kernel_regularizer=L2(0.01)

In [ ]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['acc'])
history = model.fit(sequences, y , epochs=7, validation_data=(test_seqs,y_test))

Epoch 1/7
1184/1184 [==============================] - 12s 9ms/step - loss: 0.7871 - acc: 0.6439 - val_loss: 0.5850 - val_acc: 0.7804
Epoch 2/7
1184/1184 [==============================] - 10s 9ms/step - loss: 0.5481 - acc: 0.7972 - val_loss: 0.5721 - val_acc: 0.7853
Epoch 3/7
1184/1184 [==============================] - 10s 8ms/step - loss: 0.4953 - acc: 0.8210 - val_loss: 0.5399 - val_acc: 0.7999
Epoch 4/7
1184/1184 [==============================] - 10s 8ms/step - loss: 0.4475 - acc: 0.8400 - val_loss: 0.6050 - val_acc: 0.7741
Epoch 5/7
1184/1184 [==============================] - 11s 9ms/step - loss: 0.4022 - acc: 0.8593 - val_loss: 0.5851 - val_acc: 0.7871
Epoch 6/7
1184/1184 [==============================] - 10s 9ms/step - loss: 0.3636 - acc: 0.8721 - val_loss: 0.6268 - val_acc: 0.7841
Epoch 7/7
1184/1184 [==============================] - 9s 8ms/step - loss: 0.3252 - acc: 0.8871 - val_loss: 0.6880 - val_acc: 0.7561


Top words: 15000
oov_taken: <nothing>
max_lenght: 30


ch 1/5
1184/1184 [==============================] - 15s 12ms/step - loss: 0.8330 - acc: 0.6118 - val_loss: 0.6174 - val_acc: 0.7583

Epoch 2/5
1184/1184 [==============================] - 15s 12ms/step - loss: 0.5014 - acc: 0.8184 - val_loss: 0.5618 - val_acc: 0.7956

Epoch 3/5
1184/1184 [==============================] - 13s 11ms/step - loss: 0.3805 - acc: 0.8694 - val_loss: 0.5767 - val_acc: 0.7847

Epoch 4/5
1184/1184 [==============================] - 14s 12ms/step - loss: 0.3039 - acc: 0.8965 - val_loss: 0.6100 - val_acc: 0.7747

Epoch 5/5
1184/1184 [==============================] - 13s 11ms/step - loss: 0.2520 - acc: 0.9164 - val_loss: 0.6633 - val_acc: 0.7689

Top words: 5000
oov_taken: <nothing>
max_lenght: 25

Epoch 1/5
1184/1184 [==============================] - 12s 9ms/step - loss: 0.7657 - acc: 0.6609 - val_loss: 0.6021 - val_acc: 0.7680

Epoch 2/5
1184/1184 [==============================] - 10s 9ms/step - loss: 0.5188 - acc: 0.8091 - val_loss: 0.5315 - val_acc: 0.7993

Epoch 3/5
1184/1184 [==============================] - 10s 8ms/step - loss: 0.4515 - acc: 0.8408 - val_loss: 0.5426 - val_acc: 0.8020

Epoch 4/5
1184/1184 [==============================] - 9s 8ms/step - loss: 0.3996 - acc: 0.8591 - val_loss: 0.5687 - val_acc: 0.7959

Epoch 5/5
1184/1184 [==============================] - 10s 9ms/step - loss: 0.3483 - acc: 0.8794 - val_loss: 0.6015 - val_acc: 0.7871

**Top words: 5000**
oov_taken: `<nothing> `
**max_lenght: 25**
embedding dimension: 50 (val accuracy decreased)

Epoch 1/5
1184/1184 [==============================] - 16s 13ms/step - loss: 0.7469 - acc: 0.6790 - val_loss: 0.5818 - val_acc: 0.7771

Epoch 2/5
1184/1184 [==============================] - 15s 13ms/step - loss: 0.4832 - acc: 0.8257 - val_loss: 0.5878 - val_acc: 0.7768

Epoch 3/5
1184/1184 [==============================] - 15s 13ms/step - loss: 0.3565 - acc: 0.8781 - val_loss: 0.6026 - val_acc: 0.7820

Epoch 4/5
1184/1184 [==============================] - 15s 12ms/step - loss: 0.2390 - acc: 0.9205 - val_loss: 0.6891 - val_acc: 0.7780

Epoch 5/5
1184/1184 [==============================] - 14s 12ms/step - loss: 0.1556 - acc: 0.9504 - val_loss: 0.8695 - val_acc: 0.7543